daemons/based/based_transaction.c

```diff
@@ -0,0 +1,380 @@
+              pcmk__s(client->name, "unspecified"),
+              pcmk__s(client->id, "unknown"));
+
+    if (transactions == NULL) {
+        transactions = pcmk__strkey_table(NULL, free_transaction);
+    }
+    g_hash_table_insert(transactions, client->id, g_queue_new());
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Validate that a CIB request is supported in a transaction
+ *
+ * \param[in] request  CIB request
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+validate_transaction_request(const xmlNode *request)
+{
+    const char *op = crm_element_value(request, F_CIB_OPERATION);
+    const char *host = crm_element_value(request, F_CIB_HOST);
+
+    const cib_operation_t *operation = NULL;
+    int rc = cib_get_operation(op, &operation);
+
+    rc = pcmk_legacy2rc(rc);
+    if (rc != pcmk_rc_ok) {
+        // cib_get_operation() logs error
+        return rc;
+    }
+
+    if (!pcmk_is_set(operation->flags, cib_op_attr_transaction)) {
+        crm_err("Operation '%s' is not supported in CIB transaction", op);
+        return EOPNOTSUPP;
+    }
+
+    if (!pcmk__str_eq(host, OUR_NODENAME,
+                      pcmk__str_casei|pcmk__str_null_matches)) {
+        crm_err("Operation targeting another node (%s) is not supported in CIB "
+                "transaction",
+                host);
+        return EOPNOTSUPP;
+    }
+
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Add a new CIB request to an existing transaction
+ *
+ * \param[in] client      Client whose transaction to extend
+ * \param[in] request     CIB request
+ * \param[in] privileged  If \p true, the request has write privileges
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+based_extend_transaction(pcmk__client_t *client, xmlNodePtr request,
+                         bool privileged)
+{
+    struct request_data *data = NULL;
+    GQueue *transaction = NULL;
+    int rc = pcmk_rc_ok;
+
+    CRM_ASSERT((client != NULL) && (request != NULL));
+
+    transaction = get_transaction(client);
+    if (transaction == NULL) {
+        return pcmk_rc_no_transaction;
+    }
+
+    rc = validate_transaction_request(request);
+    if (rc != pcmk_rc_ok) {
+        return rc;
+    }
+
+    data = create_request_data(client, request, privileged);
+    if (data == NULL) {
+        return ENOMEM;
+    }
+
+    g_queue_push_tail(transaction, data);
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Free a CIB client's transaction (if any) and its requests
+ *
+ * \param[in] client  Client whose transaction to discard
+ */
+void
+based_discard_transaction(const pcmk__client_t *client)
+{
+    const char *client_name = NULL;
+    const char *client_id = NULL;
+
+    CRM_ASSERT(client != NULL);
+
+    client_name = pcmk__s(client->name, "unspecified");
+    client_id = pcmk__s(client->id, "unknown");
+
+    crm_trace("Discarding transaction for client %s (%s) if one exists",
+              client_name, client_id);
```

Comment:
Don't need this message since both cases log another

Reply:
True. I added the conditional messages at the last minute because the "Discarding..." one looked weird at the end of a successful commit.